app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt

```diff
@@ -6,75 +6,85 @@
+        navController = Navigation.findNavController(activity, R.id.fragment_nav_host)
+        routeStore.routes.subscribe(this::route).addTo(compositeDisposable)
     }
 
-    private fun replaceFragment(frag: Fragment, backable: Boolean = true) {
-        val tx = activity.supportFragmentManager.beginTransaction()
-        tx.replace(R.id.root_content, frag)
-        if (backable) {
-            tx.addToBackStack(null)
-        }
-        tx.commit()
-
-        if (!backable) {
-            clearBackStack()
-        }
-    }
-
-    private fun clearBackStack() {
-        val fm = activity.supportFragmentManager
-        if (fm.backStackEntryCount > 0) {
-            val base = fm.getBackStackEntryAt(0)
-            fm.popBackStackImmediate(base.id, FragmentManager.POP_BACK_STACK_INCLUSIVE)
+    private fun replaceFragment(@IdRes destinationId: Int, args: Bundle? = null) {
+        val from = navController.currentDestination?.id ?: return
+        if (from == destinationId && args == null) {
+            // No point in navigating if nothing has changed.
+            return
         }
+        val transition = findTransitionId(from, destinationId) ?: destinationId
+        navController.navigate(transition, args)
     }
 
     private fun route(action: RouteAction) {
         when (action) {
-            is RouteAction.Welcome -> replaceFragment(welcome, false)
-            is RouteAction.Login -> replaceFragment(login)
-            is RouteAction.ItemList -> replaceFragment(itemList, false)
-            is RouteAction.SettingList -> replaceFragment(settingList)
-            is RouteAction.LockScreen -> replaceFragment(lock, false)
-            is RouteAction.Filter -> replaceFragment(filter)
+            is RouteAction.Welcome -> replaceFragment(R.id.fragment_welcome)
+            is RouteAction.Login -> replaceFragment(R.id.fragment_fxa_login)
+            is RouteAction.ItemList -> replaceFragment(R.id.fragment_item_list)
+            is RouteAction.SettingList -> replaceFragment(R.id.fragment_setting)
+            is RouteAction.LockScreen -> replaceFragment(R.id.fragment_locked)
+            is RouteAction.Filter -> replaceFragment(R.id.fragment_filter)
             is RouteAction.ItemDetail -> {
-                itemDetail.itemId = action.id
-                replaceFragment(itemDetail)
+                // Possibly overkill for passing a single id string,
+                // but it's typesafe™.
+                val bundle = ItemDetailFragmentArgs.Builder()
+                        .setItemId(action.id)
+                        .build()
+                        .toBundle()
+                replaceFragment(R.id.fragment_item_detail, bundle)
             }
-            is RouteAction.Back -> activity.supportFragmentManager.popBackStack()
+            is RouteAction.Back -> navController.popBackStack()
         }
     }
+
+    private fun findTransitionId(@IdRes from: Int, @IdRes to: Int): Int? {
+        when (from) {
```

Comment:
is there a cleaner way to encode the `from` and `to` transitions than this nested `when` structure?

Reply:
I'm thinking `when(from, to)` and then we can construct the pairs as we construct the actions.

it's mostly the nesting that bugs me here.

Reply:
Oh yes, that's a good idea. This is the code I'm least in love with.

Reply:
Not wild about this, but this is more in keeping with what you were after.

```kt
when (Pair(from, to)) {
    Pair(R.id.fragment_welcome, R.id.fragment_fxa_login) -> return R.id.action_welcome_to_fxaLogin
    …
}
```